src/esp/gfx/RenderCamera.cpp

```diff
@@ -212,8 +213,11 @@ esp::geo::Ray RenderCamera::unproject(const Mn::Vector2i& viewportPosition) {
   ray.direction =
       ((object().absoluteTransformationMatrix() * projectionMatrix().inverted())
            .transformPoint(normalizedPos) -
-       ray.origin)
-          .normalized();
+       ray.origin) /
+      1000.0;
```

Comment:
Why 1000?  Should this be a defined constant somewhere?

Reply:
Considering the intent of this change, perhaps some documentation regarding rescaling the normalized camera ray to fit the desired use would be more appropriate?

Reply:
Great question, this is the far plane distance.
I updated the code and docs such that the scaling is on the user side.
Also added a test to validate this.

Reply:
how about we get the far distance by doing this right there? @mosra does this look appropriate?

```
auto zFar = Frustum::fromMatrix(projectionMatrix()*cameraMatrix()).far();
auto dist = zFar(3)/zFar.xyz().length();

```

Reply:
The `Frustum` class is unnecessary overhead -- it calculates also all other planes and you need just one of them, and only its distance.

The "most obvious" way to calculate the distance of the far plane is by doing an inverse transform of a point lying on the far plane (i.e., calculating what point gets projected to `(0, 0, 1)`). Something like this:

```cpp
projectionMatrix().inverted().transformPoint(Vector3::zAxis(1.0f)).z()
```

But that's imprecise and not fast either due to the inversion. After looking at the [Matrix4::perspectiveProjection()](https://doc.magnum.graphics/magnum/classMagnum_1_1Math_1_1Matrix4.html#a6475b3ef155c9142b890c8133504ae9b) docs and a bit of math suffering I think the equation is like this:

```cpp
near = projection[3][2]/(projection[2][2] - 1.0f);
far = projection[3][2]/(1.0f + projection[2][2]);
```

I could add those directly on the Matrix4 class I think, just need to do a bit of testing and correctly handle infinite far plane.

Reply:
Thanks @mosra ! Do we not need the product of the projection & camera transforms for the far plane calculation?

Reply:
Yes -- the equations I have give you the distance to where far Z plane is relative to the camera. So then if you transform the camera somewhere, you don't need to do the complicated extraction of the far plane again, only transforming that "far vector" with the camera matrix. I think. My math is rusty.

```cpp
// Negative Z is forward
Vector3 cameraRelativeZIsHere = cameraMatrix().transformPoint(Vector3::zAxis(-far));
```